File: app/src/main/java/com/example/pam_project/features/categories/edit/EditCategoryPresenter.java
package com.example.pam_project.features.categories.edit;

import com.example.pam_project.features.categories.list.CategoryInformation;
import com.example.pam_project.repositories.categories.CategoriesRepository;
import com.example.pam_project.utils.schedulers.SchedulerProvider;

import java.lang.ref.WeakReference;

import io.reactivex.disposables.Disposable;

public class EditCategoryPresenter {
    private final long categoryId;
    private final SchedulerProvider provider;
    private final CategoriesRepository repository;
    private final WeakReference<EditCategoryView> view;
    private Disposable fetchCategoryDisposable;
    private Disposable updateCategoryDisposable;
    private Disposable deleteCategoryDisposable;

    public EditCategoryPresenter(final long categoryId, final SchedulerProvider provider, final CategoriesRepository repository,
                                 final EditCategoryView view) {
        this.categoryId = categoryId;
        this.provider = provider;
        this.repository = repository;
        this.view = new WeakReference<>(view);
    }

    public void onViewAttached() {
        fetchCategoryDisposable = repository.getCategory(categoryId)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe(this::onCategoryRetrieved, this::onCategoryRetrievedError);
    }

    private void onCategoryRetrieved(final CategoryInformation model) {
        if (view.get() != null) {
            view.get().bindCategory(model);
        }
    }

    private void onCategoryRetrievedError(final Throwable throwable) {
        if (view.get() != null) {
            view.get().onCategoryRetrievedError();
        }
    }

    public void updateCategory(final String name, final String color) {
        updateCategoryDisposable = repository.updateCategory(categoryId, name, color)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe(()->{},this::onCategoryUpdateError);
    }

    private void onCategoryUpdateError(final Throwable throwable) {
        if (view.get() != null) {
            view.get().onCategoryUpdateError();
        }
    }

    public void deleteCategory(final long id) {
        deleteCategoryDisposable = repository.deleteCategory(id)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe(this::onCategoryDeleted, this::onCategoryDeletedError);
    }

    private void onCategoryDeleted() {
        if (view.get() != null) {
            view.get().onCategoryDelete();
        }
    }

    private void onCategoryDeletedError(final Throwable throwable) {
        if (view.get() != null) {
            view.get().onCategoryDeletedError();
        }
    }

    public void onDeletePressed() {
        if (view.get() != null) {
            view.get().showDeleteDialog();
        }
    }

    public void onViewDetached() {
        if (fetchCategoryDisposable != null)
            fetchCategoryDisposable.dispose();
        if (updateCategoryDisposable != null)
            updateCategoryDisposable.dispose();
        if (deleteCategoryDisposable != null)
            deleteCategoryDisposable.dispose();
    }

}

File: app/src/main/java/com/example/pam_project/features/categories/edit/EditCategoryPresenter.kt
package com.example.pam_project.features.categories.edit

import com.example.pam_project.features.categories.list.CategoryInformation
import com.example.pam_project.repositories.categories.CategoriesRepository
import com.example.pam_project.utils.schedulers.SchedulerProvider
import io.reactivex.disposables.Disposable
import java.lang.ref.WeakReference

class EditCategoryPresenter(private val categoryId: Long, private val provider: SchedulerProvider?,
                            private val repository: CategoriesRepository?, view: EditCategoryView?) {
    private val view: WeakReference<EditCategoryView?> = WeakReference(view)
    private var fetchCategoryDisposable: Disposable? = null
    private var updateCategoryDisposable: Disposable? = null
    private var deleteCategoryDisposable: Disposable? = null

    fun onViewAttached() {
        fetchCategoryDisposable = repository!!.getCategory(categoryId)
                .subscribeOn(provider!!.computation())
                .observeOn(provider.ui())
                .subscribe({ model: CategoryInformation? -> onCategoryRetrieved(model) }) { throwable: Throwable
                    -> onCategoryRetrievedError(throwable) }
    }

    private fun onCategoryRetrieved(model: CategoryInformation?) {
        if (view.get() != null) {
            view.get()!!.bindCategory(model)
        }
    }

    private fun onCategoryRetrievedError(throwable: Throwable) {
        if (view.get() != null) {
            view.get()!!.onCategoryRetrievedError()
        }
    }

    fun updateCategory(name: String, color: String) {
        updateCategoryDisposable = repository!!.updateCategory(categoryId, name, color)
                .subscribeOn(provider!!.computation())
                .observeOn(provider.ui())
                .subscribe({}) { throwable: Throwable -> onCategoryUpdateError(throwable) }
    }

    private fun onCategoryUpdateError(throwable: Throwable) {
        if (view.get() != null) {
            view.get()!!.onCategoryUpdateError()
        }
    }

    fun deleteCategory(id: Long) {
        deleteCategoryDisposable = repository!!.deleteCategory(id)
                .subscribeOn(provider!!.computation())
                .observeOn(provider.ui())
                .subscribe({ onCategoryDeleted() }) { throwable: Throwable -> onCategoryDeletedError(throwable) }
    }

    private fun onCategoryDeleted() {
        if (view.get() != null) {
            view.get()!!.onCategoryDelete()
        }
    }

    private fun onCategoryDeletedError(throwable: Throwable) {
        if (view.get() != null) {
            view.get()!!.onCategoryDeletedError()
        }
    }

    fun onDeletePressed() {
        if (view.get() != null) {
            view.get()!!.showDeleteDialog()
        }
    }

    fun onViewDetached() {
        if (fetchCategoryDisposable != null) fetchCategoryDisposable!!.dispose()
        if (updateCategoryDisposable != null) updateCategoryDisposable!!.dispose()
        if (deleteCategoryDisposable != null) deleteCategoryDisposable!!.dispose()
    }

}